src/piracy.jl

```diff
@@ -77,71 +78,118 @@ is_foreign(@nospecialize(x), pkg::Base.PkgId) = is_foreign(typeof(x), pkg)
 
 is_foreign_module(mod::Module, pkg::Base.PkgId) = Base.PkgId(mod) != pkg
 
-function is_foreign(@nospecialize(T::DataType), pkg::Base.PkgId)
+function is_foreign(
+    @nospecialize(T::DataType),
+    pkg::Base.PkgId;
+    treat_as_own_type::Vector{<:Type} = Type[],
+)
     params = T.parameters
     # For Type{Foo}, we consider it to originate from the same as Foo
     C = getfield(parentmodule(T), nameof(T))
     if C === Type
         @assert length(params) == 1
-        return is_foreign(first(params), pkg)
+        return is_foreign(first(params), pkg; treat_as_own_type = treat_as_own_type)
     else
         # Both the type itself and all of its parameters must be foreign
-        return is_foreign_module(parentmodule(T), pkg) && all(params) do param
-            is_foreign(param, pkg)
-        end
+        return (!(C in treat_as_own_type) && is_foreign_module(parentmodule(T), pkg)) &&
+               all(
+            param -> is_foreign(param, pkg; treat_as_own_type = treat_as_own_type),
+            params,
+        )
     end
 end
 
-function is_foreign(@nospecialize(U::UnionAll), pkg::Base.PkgId)
+function is_foreign(
+    @nospecialize(U::UnionAll),
+    pkg::Base.PkgId;
+    treat_as_own_type::Vector{<:Type} = Type[],
+)
     # We do not consider extending Set{T} to be piracy, if T is not foreign.
     # Extending it goes against Julia style, but it's not piracy IIUC.
-    is_foreign(U.body, pkg) && is_foreign(U.var, pkg)
+    is_foreign(U.body, pkg; treat_as_own_type = treat_as_own_type) &&
+        is_foreign(U.var, pkg; treat_as_own_type = treat_as_own_type)
 end
 
-is_foreign(@nospecialize(T::TypeVar), pkg::Base.PkgId) = is_foreign(T.ub, pkg)
+is_foreign(
+    @nospecialize(T::TypeVar),
+    pkg::Base.PkgId;
+    treat_as_own_type::Vector{<:Type} = Type[],
+) = is_foreign(T.ub, pkg; treat_as_own_type = treat_as_own_type)
 
 # Before 1.7, Vararg was a UnionAll, so the UnionAll method will work
 @static if VERSION >= v"1.7"
-    is_foreign(@nospecialize(T::Core.TypeofVararg), pkg::Base.PkgId) = is_foreign(T.T, pkg)
+    is_foreign(
+        @nospecialize(T::Core.TypeofVararg),
+        pkg::Base.PkgId;
+        treat_as_own_type::Vector{<:Type} = Type[],
+    ) = is_foreign(T.T, pkg; treat_as_own_type = treat_as_own_type)
 end
 
-function is_foreign(@nospecialize(U::Union), pkg::Base.PkgId)
+function is_foreign(
+    @nospecialize(U::Union),
+    pkg::Base.PkgId;
+    treat_as_own_type::Vector{<:Type} = Type[],
+)
     # Even if Foo is local, overloading f(::Union{Foo, Int}) with foreign f
     # is piracy.
-    any(T -> is_foreign(T, pkg), Base.uniontypes(U))
+    any(T -> is_foreign(T, pkg; treat_as_own_type = treat_as_own_type), Base.uniontypes(U))
 end
 
-function is_foreign_method(@nospecialize(U::Union), pkg::Base.PkgId)
+function is_foreign_method(
+    @nospecialize(U::Union),
+    pkg::Base.PkgId;
+    treat_as_own_type::Vector{<:Type} = Type[],
+    treat_as_own_func::Vector{<:Function} = Function[],
```

Comment:
What do we gain by having two arguments? It seems fine to mix both types and functions in a single list, and arguably might be easier to use and read?

Reply:
I guess the only downside to this is that julia cannot specialize in the parameters anymore. But this should not make that big of a difference. Thus I merged the two kwargs in 9e0a05b7841d5fa00585b93828682b810f55f114.